✨ Display sync status next to the logout button

pdrj opened the pull request:
## What's this?
This displays sync status next to the logout button in settings.

### Relationships

Closes #4907

## Why do we want this?
iOS has it and it gives users more info about their state.

## How is it done?
Moving an observable from iOS solution to the view model and then some layout work.

### Why not another way?
This seems pretty clean.

## Review hints
Trigger sync, go to settings quickly. You should see the following.

![Screenshot 2019-09-27 at 11 15 46](https://user-images.githubusercontent.com/53510080/65763588-0b874f00-e124-11e9-881f-bcde1bf36b47.png)
![Screenshot 2019-09-27 at 11 15 57](https://user-images.githubusercontent.com/53510080/65763590-0b874f00-e124-11e9-9f64-73f29b8861e6.png)
![Screenshot 2019-09-27 at 11 17 03](https://user-images.githubusercontent.com/53510080/65763593-0b874f00-e124-11e9-8eb1-056ecc90d4a0.png)
![Screenshot 2019-09-27 at 11 16 41](https://user-images.githubusercontent.com/53510080/65763595-0b874f00-e124-11e9-8ddc-3efd0a28796c.png)

Try to logout as well.

If we want to add some fancy design, I'd say we can do it later and now merge it as it is. @maritoggl please have a look at the above screenshots and say if you're OK with how it looks. It is done in a similar way on iOS.

## :squid: Permissions
Me,

dshr commented on Toggl.Core.UI/ViewModels/Settings/SettingsViewModel.cs:
```diff
@@ -54,13 +54,21 @@ public sealed class SettingsViewModel : ViewModel
         public bool CalendarSettingsEnabled => onboardingStorage.CompletedCalendarOnboarding();
         public string Version => $"{platformInfo.Version} ({platformInfo.BuildNumber})";
+        
+        public enum SyncStatus
```
Perhaps could extract this into a separate file in `Toggl.Core.Sync`, like `SyncState.cs`? The iOS side of things looks great otherwise.

pdrj replied:
Yeah I thought about it, but then we'd have `SyncState` and `SyncStatus` there - maybe a bit confusing? 🤷‍♂ I also thought about just reusing the one from `Core.Sync` but it doesn't quite match the values needed on the UI.

maritoggl commented:
I’m ok with this approach, but do we cover all instances with just “Synced” and “Syncing”? I think we should have a status that reflects there are unsynced entries as well as those are the most important entries in the context of logging out

pdrj commented:
@maritoggl I'm doing exactly what iOS is - there is 'synced', 'syncing' and 'logging out'. It would be useful to have what you propose, but maybe we should make a separate issue for that? What do you think?

maritoggl commented:
I agree, no need to extend the scope of this PR.

pdrj commented:
Thanks Mari. Only waiting for @toggl/androids review then.

Odrakir commented on Toggl.Core.UI/ViewModels/Settings/SettingsViewModel.cs:
```diff
@@ -25,6 +25,7 @@
 using Toggl.Shared.Extensions;
 using Toggl.Storage.Settings;
 using Xamarin.Essentials;
+using PresentableSyncStatus = Toggl.Core.Sync.PresentableSyncStatus;
```
I think you don't need this line, given that you are importing `Toggl.Core.Sync`

Odrakir commented on Toggl.iOS/Cells/Settings/SettingsSyncCell.cs:
```diff
@@ -7,10 +7,11 @@
 using Toggl.Shared;
 using UIKit;
 using Colors = Toggl.Core.UI.Helper.Colors;
+using PresentableSyncStatus = Toggl.Core.Sync.PresentableSyncStatus;
```
This is probably me not understanding the way things are done in C#, but isn't it clearer (more consistent) to just do `using Toggl.Core.Sync;'?

kornelijepetak replied:
No, @Odrakir, your understanding of C# is on point.
The only reason why one would use an alias in this case is if `using Toggl.Core.Sync` would make visible some other type that would clash with another namespace. Which in this situation is not the case.

kornelijepetak commented on Toggl.Core.UI/ViewModels/Settings/SettingsViewModel.cs:
```diff
@@ -207,6 +208,16 @@ public sealed class SettingsViewModel : ViewModel
+                IsSynced,
+                IsRunningSync,
+                LoggingOut.SelectValue(true).StartWith(false),
+                (synced, syncing, loggingOut) =>
```
Please extract this into a separate method to respect the levels of abstractions, especially when you are not utilizing closures.

Please do not write returns in the same line as `if`s.

kornelijepetak commented on Toggl.Droid/Fragments/SettingsFragment.ViewBinds.cs:
```diff
@@ -79,7 +88,8 @@ protected override void InitializeViews(View fragmentView)
       
-            logoutView = fragmentView.FindViewById<TextView>(Resource.Id.SettingsLogoutButton);
+            txtLogout = fragmentView.FindViewById<TextView>(Resource.Id.TxtSettingsLogout);
+            logoutView = fragmentView.FindViewById<View>(Resource.Id.SettingsLogoutButton);
```
```suggestion
            logoutView = fragmentView.FindViewById(Resource.Id.SettingsLogoutButton);
```

kornelijepetak commented on Toggl.Droid/Fragments/SettingsFragment.ViewBinds.cs:
```diff
@@ -121,8 +131,16 @@ protected override void InitializeViews(View fragmentView)
-
-            logoutView.Text = Shared.Resources.SignOutOfToggl;
+            
+            syncStateSyncedHolder = fragmentView.FindViewById<View>(Resource.Id.SyncStateSyncedHolder);
```
```suggestion
            syncStateSyncedHolder = fragmentView.FindViewById(Resource.Id.SyncStateSyncedHolder);
```

kornelijepetak commented on Toggl.Droid/Fragments/SettingsFragment.ViewBinds.cs:
```diff
@@ -121,8 +131,16 @@ protected override void InitializeViews(View fragmentView)
-            logoutView.Text = Shared.Resources.SignOutOfToggl;
+            
+            syncStateSyncedHolder = fragmentView.FindViewById<View>(Resource.Id.SyncStateSyncedHolder);
+            syncStateInProgressHolder = fragmentView.FindViewById<View>(Resource.Id.SyncStateInProgressHolder);
```
```suggestion
            syncStateInProgressHolder = fragmentView.FindViewById(Resource.Id.SyncStateInProgressHolder);
```

kornelijepetak commented on Toggl.Droid/Fragments/SettingsFragment.cs:
```diff
@@ -164,6 +167,20 @@ public override void OnViewCreated(View view, Bundle savedInstanceState)
+
+            txtStateInProgress.Text = status == PresentableSyncStatus.Syncing ? Syncing : LoggingOutSecurely;
+
+            syncStateViews.TryGetValue(status, out var visibleView);
```
If you are not using the `bool` return value to check for existance of the key, please use the dictionary access syntax.
```suggestion
            var visibleView = syncStateViews[status];
```

kornelijepetak commented on Toggl.iOS/Cells/Settings/SettingsSyncCell.cs:
```diff
@@ -46,19 +47,19 @@ protected override void UpdateView()
             switch (Item.CustomValue)
             {
-                case SyncStatus.Synced:
+                case PresentableSyncStatus.Synced:
```
You can avoid writing these with
`using static Toggl.Core.Sync.PresentableSyncStatus;`

Then just

`case Synced:`
`case Syncing:`
...

pdrj replied:
If I do this and remove `using Toggl.Core.Sync`, this line no longer compiles: `    public partial class SettingsSyncCell : BaseTableViewCell<CustomRow<PresentableSyncStatus>>`. Am I doing something wrong?

kornelijepetak replied:
Yeah, just add a `using static Toggl.Core.Sync.PresentableSyncStatus` without removing the `using Toggl.Core.Sync`.

`using static` just tells the compiler that for this file, the enum elements do not need to be fully qualified.

semanticer commented on Toggl.Droid/Resources/layout/SettingsFragment.axml:
```diff
@@ -613,18 +613,63 @@
+                    android:layout_height="32dp"
+                    android:layout_width="match_parent"
+                    android:layout_marginBottom="2dp"/>
+                <FrameLayout
```
I don't love how deep the layout hierarchy is but it's consistent with the rest of the layout and it looks like this is how we roll for now :-(

pdrj commented:
@kornelijepetak - I addressed all your comments except the one about static `using`. Please let me know what would you like to be done there, WRT to my reply to your comment.
